src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+ duration of 1 and unique start times, ordered to match the flow of the program. Layering compresses the timeline by
+ assuming that some operations might be executed simultaneously while allowing for different operations to have various
+ durations.
+
+_Definition_: Provided a valid _time_ function for the program a ___layer of duration N at time T, denoted as L(T,N),___
+ is a subset of operations in the program such that all of these operations have start time greater or equal _T_ and
+ finish time less than _T + N_. The program is ___layered___ if all gates in it are partitioned into layers, that don't
+ overlap in time. The union of all qubits that are involved in operations of a given layer, will be denoted _Qubits(T,N)_.
+
+A sequential program can be trivially layered such that each layer contains exactly one operation. Notice, that the
+ definition of layer doesn't require the gates to be executed _in parallel_. For example, all gates in a fully sequential
+ program can be also placed into a single layer L(0, infinity). Some gates might be considered to be very cheap and take
+ zero time to execute, those gates can be added to a layer even if they act on the same qubit another gate in this layer
+ is acting on and have to be executed sequentially within the layer.
+
+_Definition_: A ___global barrier___ is any operation that acts on _all_ currently allocated qubits. We'll provide means
+ for the clients to inject global barriers equivalent to an identity operator. This will allow the clients to enforce a
+ particular layering structure (because no later operation can sink below the barrier).
+
+The global barriers will be implemented as calls to `__quantum__qis__global_barrier` function. No additional support from
+ the compiler should be needed, as the user can define their own intrinsic to represent the barrier and map it to the above
+ runtime function via targets.qs file. The user can choose duration of a barrier which would affect start time of the
+ following layers but no operations will be added to a barrier, independent of its width.
+
+### The Resource Tracer's Layering Algorithm ###
+
+As the tracer is executing a sequential quantum program, it will compute a time function and corresponding layering
+ using the _conceptual_ algorithm, described below (aka "tetris algorithm"). The actual implementation of layering might
+ be done differently, as long as the resulting layering is the same as if running the conceptual algorithm.
+
+1. The tracer will have a setting for preferred layer duration: P.
+1. The first encountered operation of __non-zero__ duration N is added into layer L(0, max(P,N)). The value
+ for _conditional barrier_ is set to 0.
+1. When conditional callback is encountered, the layer L(t,N) of the measurement that produced the result the conditional
+ is dependent on, is looked up and the _conditional barrier_ is set to _t + N_. At the end of the conditional scope the

```

Comment:
> When conditional callback is encountered, the layer L(t,N) of the measurement that produced the result the conditional is dependent on, is looked up and the _conditional barrier_ is set to _t + N_. At the end of the conditional scope ...

Here is how I interpret it:
> The quantum program can contain conditionals. Before executing a conditional, the _conditional callback_ (provided by the Tracer, described earlier in <where?>) is invoked. The _conditional callback_ knows that the conditional depends on the result produced during the measurement done earlier in layer L(t,N). The _conditional callback_ finds that layer L(t,N) and places the _conditional barrier_ at a time stamp t + N (in the end of that layer). Then the _conditional callback_ returns the control back to the quantum program, the conditional is executed, and the block of code inside of the conditional is affected by the _conditional barrier_ that has just been set (i.e. the set of layers, available for the block of code inside of the conditional, is limited by the latest placed _conditional barrier_). At the end of the conditional scope the _conditional barrier_ is placed back to the time stamp of the outer conditional (or to 0 if there is no outer conditional), i.e. the result of the previous invocation of _conditional callback_ is restored (in case of nested conditionals).